### PR TITLE
[MachineScheduler] Add option to control reordering for store/load clustering

### PR DESCRIPTION
**Pull request by asb**
Reordering based on the sort order of the MemOpInfo array was disabled in <https://reviews.llvm.org/D72706>. However, it's not clear this is desirable for al targets. It also makes it more difficult to compare the incremental benefit of enabling load clustering in the selectiondag scheduler as well was the machinescheduler, as the sdag scheduler does seem to allow this reordering.

This patch adds a parameter that can control the behaviour on a per-target basis.

Split out from #73789.

### CODE REVIEW

**Comment by asb** [comment]
Ping.

**Comment by jayfoad** [comment]
Are you saying that reordering the instructions to match the order in MemOpInfo is positively desirable in some cases? I only disabled it because it seemed completely arbitrary to me.

I think it would be even better if we could find some way to schedule the instructions adjacently without enforcing any particular order, but I don't see any way to do this just by adding edges to the DAG.

**Comment by asb** [comment]
> Are you saying that reordering the instructions to match the order in MemOpInfo is positively desirable in some cases? I only disabled it because it seemed completely arbitrary to me.

It's definitely somewhat arbitrary. I don't see it being harmful though, and it's attractive to me that the same sorting is used for clustering in the MachineScheduler as for the SelectionDAG (which also sorts in increasing order of offset in ScheduleDAGSDNodes::ClusterNeighbouringLoads). As a reader of the generated assembly, if the order of neighbouring loads is unimportant then I also have a slight preference for them being in sorted order as that's clearly arbitrary, while unsorted has the false appearance of being meaningful.

I'd say matching the SDag scheduler is the main one that motivated bothering to patch this - having the same order makes it easier to compare what missed opportunities / differences there are when enabling load clustering there as well.

**Comment by jayfoad** [comment]
> > Are you saying that reordering the instructions to match the order in MemOpInfo is positively desirable in some cases? I only disabled it because it seemed completely arbitrary to me.
> 
> It's definitely somewhat arbitrary. I don't see it being harmful though, and it's attractive to me that the same sorting is used for clustering in the MachineScheduler as for the SelectionDAG (which also sorts in increasing order of offset in ScheduleDAGSDNodes::ClusterNeighbouringLoads). As a reader of the generated assembly, if the order of neighbouring loads is unimportant then I also have a slight preference for them being in sorted order as that's clearly arbitrary, while unsorted has the false appearance of being meaningful.
> 
> I'd say matching the SDag scheduler is the main one that motivated bothering to patch this - having the same order makes it easier to compare what missed opportunities / differences there are when enabling load clustering there as well.

I did not know about ScheduleDAGSDNodes::ClusterNeighbouringLoads. I don't have any objection to this patch.

File: llvm/lib/Target/RISCV/RISCVTargetMachine.cpp [hunk 1]
```diff
@@ -348,7 +348,7 @@ class RISCVPassConfig : public TargetPassConfig {
     ScheduleDAGMILive *DAG = nullptr;
     if (EnableMISchedLoadClustering) {
       DAG = createGenericSchedLive(C);
-      DAG->addMutation(createLoadClusterDAGMutation(DAG->TII, DAG->TRI));
+      DAG->addMutation(createLoadClusterDAGMutation(DAG->TII, DAG->TRI, true));
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:351]: add a comment to label `true` here

**Comment by asb** [comment]
Ping.

**Review by preames — APPROVED** [review]
LGTM

**Pull request merged**